src/libpmemobj/heap.c

```diff
@@ -1046,6 +1070,62 @@ heap_extend(struct palloc_heap *heap, struct bucket *b, size_t size)
 	return 1;
 }
 
+/*
+ * heap_zone_fixup_size -- mask of potentially unreachable memory from
+ *	pools affected by a size alignment bug.
+ *
+ * This will run only for pools that have a mismatch between the on-media
+ * size index and the correct runtime-calculated size index.
+ */
+static void
+heap_zone_fixup_size(struct palloc_heap *heap, uint32_t zone_id)
+{
+	struct zone *z = ZID_TO_ZONE(heap->layout, zone_id);
+	uint32_t size_idx_old = z->header.size_idx;
+	size_t size_idx = zone_calc_size_idx(zone_id, heap->rt->nzones,
+		*heap->sizep);
+
+	for (uint32_t i = 0; i < size_idx_old; ) {
+		struct chunk_header *hdr = &z->chunk_headers[i];
+		ASSERT(hdr->size_idx != 0);
+
+		uint32_t next_s = i + hdr->size_idx;
+		if (next_s > size_idx) {
+			if (hdr->type == CHUNK_TYPE_FREE) {
+				/* just re-init the zone */
+				heap_zone_init(heap, zone_id, i);
+			} else if (hdr->type == CHUNK_TYPE_RUN) {
+				/* mark the remaining bits as allocated */
+				struct memory_block m = MEMORY_BLOCK_NONE;
+				m.chunk_id = i;
+				m.zone_id = zone_id;
+				memblock_rebuild_state(heap, &m);
+				struct run_bitmap b;
+				m.m_ops->get_bitmap(&m, &b);
+				/*
+				 * We need to prevent allocation of the last
+				 * ~4 kilobytes of the last run. To do that,
+				 * we will set all bits to 1 in the last two
+				 * values (so total of 128 bits).
+				 * This doesn't really have to be atomic
+				 * because for affected pools the code will
+				 * always have to go through this path.
+				 */
+				unsigned values = MIN(b.nvalues, 2);
+				for (unsigned i = b.nvalues - values;
+					i < b.nvalues; ++i) {
+					pmemops_memset(&heap->p_ops,
```

Comment:
what if this memory was already allocated? is the user pool corrupted? Shouldn't we fail in this case?